Add Lookup Row Index to Refl View

cailafinn opened the pull request:
**Description of work.**

The reflectometry scientists wanted a way to see which of the "per-angle defaults" (probably needs renaming now it takes the title of the group into account too) were used for each of the runs in the runs table. 
This is the final element of that work, adding the functionality to the view so that the user can see what's going on under the hood. 
When the user has a number of rows in the runs table and some matching rows in the experiment lookup table, the user should be able to see which of the rows in the lookup table were used by looking at the final column on the runs table. 

**To test:**

1. Open the reflectometry GUI
2. Open the "Batch" menu and load the `.json` file that appears in the ISIS Sample Data. 
3. Click the arrows next to the loaded groups to open the drop-downs
4. Check the names of the groups and the angles for each of the rows.
5. Expand the GUI, you should see a column called Lookup Index that says "1" for each row (We're using 1 based indexes. I don't like it either, but it's better for the users)
6. Navigate to the Experiment Settings Tab
7. The 1 refers to the single, empty line currently present in the "Per-angle Defaults" table
8. The line being empty in the Angle and Title columns means that this is a Wildcard Row, and will match anything. 
9. Add a row or two to match some of the runs on the runs table. (Note: Matches happen from most to least specific; Matching title & angle -> Matching angle -> wildcard)
10. You'll need to see if the processing is working, so put in values like 0.01, 0.02, etc. into the dQ/Q column. 
11. Navigate back to the runs table
12. The "Lookup Index" column should now contain the index of the row you expect it to match in the experiment settings tab
13. Change group names and angles and check that the lookup index column is updated correctly. 
14. Click process and check that the correct dQ/Q values were taken from the experiment settings tab

Fixes #33626 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

Commit:
Add lookup index column to view

Added the column to the view. It isn't updated or useful yet, but it's
visible.

RE #33626

Commit:
Update view when row index updated

RE #33626

Commit:
Rename RowOutputsChanged now triggers are different

RE #33626

Commit:
Refactor method only called after ModelChanged

Calls to notifyRowStateChanged are only performed after a call to
notifyRowModelChanged in the BatchPresenter. Move the call about a state
changed to inside the ModelChanged method of the RunsPresenter.

RE #33626

Commit:
Add release notes

RE #33626

robertapplin reviewed (changes requested):
It works as described and the code looks good. I just had a few minor suggestions

Commit:
Tidy index incrementation code

Using the ternary operator makes this much nicer. Also realised the
implementation is in an anon namespace, so doesn't need to be in the
header.

RE #33626

ConorMFinn commented on qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/JobsViewUpdater.cpp:
```diff
@@ -35,7 +35,18 @@ MantidWidgets::Batch::Cell qRangeCellOrDefault(RangeInQ const &qRangeInput, Rang
+    lookupIndex = lookupIndex.get() + 1;
+  } else {
+    lookupIndex = boost::none;
+  }
```
Much nicer looking with it, thanks

ConorMFinn commented on qt/scientific_interfaces/ISISReflectometry/GUI/RunsTable/JobsViewUpdater.cpp:
```diff
@@ -35,7 +35,18 @@ MantidWidgets::Batch::Cell qRangeCellOrDefault(RangeInQ const &qRangeInput, Rang
   return result;
 }
 
+boost::optional<size_t> incrementIndex(const Row &row) {
```
Non-member functions cannot be `const`, as the function `const` qualifier dictates it will not change any member variables, but doesn't have any members it could even change

robertapplin replied:
Ah yes good point, think I got confused as I saw it in the header file

ConorMFinn replied:
No worries, you made me notice it didn't need to be in the header anyway

robertapplin reviewed (approved):
Thanks for the changes, it looks good to be merged 👍

gemmaguest commented on qt/scientific_interfaces/ISISReflectometry/GUI/Common/Decoder.cpp:
```diff
@@ -286,7 +287,7 @@ void Decoder::decodeRunsTable(QtRunsTableView *gui, ReductionJobs *redJobs, Runs
   if (m_projectSave) {
     // Apply styling and restore completed state for output range values
-    presenter->notifyRowOutputsChanged();
+    presenter->notifyRowModelChanged();
```
Does `notifyRowModelChanged` need moving outside this `if` statement now? Previously this was only applicable when doing a save of the whole project (including output workspaces, hence the processed-state could be saved and restored), but I think this will also be needed to restore the index when loading a batch file. Although I'm not sure - does it currently work?

ConorMFinn replied:
It currently works as-is, loading a batch or a project keeps the lookup index how I'd expect them to be
